# CSE185-SP24-Motif-Finding-by-Gibbs-Sampling
## Implementation
Starting with a random selection of k-mers from the sequences, the Gibbs sampler we implemented iteratively refine this selection to converge on the most probable motifs. In each iteration, a profile matrix is created from the current set of motifs, excluding one sequence. This matrix represents the probabilities of each nucleotide at each position. Using this profile, a new k-mer is probabilistically selected for the excluded sequence based on its likelihood given the profile. This process is repeated, gradually reconfines the motifs by the probabilistic framework. By running the algorithm multiple times and using a background model to calculate p-values, the implementation aims to identify statistically significant motifs that are most representative of the biological signals.
## Usage
This program is not yet developed to be a command-line executable utility. Therefore, to run this program, the myutils.py script has to be executed conventionally by using python myutils.py. To run the program as intended, the file location needs to be updated within the myutils.py script at line 105: fasta_file = "location/of/peakdatafile.fa". Parameters such as k, t, and n, corresponding to motif length, number of sequences, and number of iterations, can be varied to explore the resulting motifs.
